### ISSUE TITLE
Manual hyperparameter optimization

### ISSUE THREAD

**Issue by BSonya** [MIC-DKFZ/nnUNet#1580]
Hi. 
I'm using nnUNetv2 for the Hecktor dataset on my linux machine. I want to use `nnUNetTrainerDiceTopK10Loss` as trainer class while also setting up num_epochs = 50/100. I have tried to use two trainer classes simultaneously like this  **`nUNetv2_train 95 3d_fullres all -tr nnUNetTrainerDiceTopK10Loss -tr nnUNetTrainer_20epochs --npz`**  but it doesn't seem to use the loss as specified. 
Is there any alternate way of manual hyperparameter optimization? like num_epochs/ batch_size etc?

**Comment by MachtaYassine** [comment]
```
class nnUNetTrainerTopk10Loss_50epochs(nnUNetTrainer):
    def __init__(self, plans: dict, configuration: str, fold: int, dataset_json: dict, unpack_dataset: bool = True,
                 device: torch.device = torch.device('cuda')):
        super().__init__(plans, configuration, fold, dataset_json, unpack_dataset, device)
        self.num_epochs = 50
    def _build_loss(self):
        assert not self.label_manager.has_regions, 'regions not supported by this trainer'
        loss = TopKLoss(ignore_index=self.label_manager.ignore_label if self.label_manager.has_ignore_label else -100,
                        k=10)
        deep_supervision_scales = self._get_deep_supervision_scales()
        # we give each output a weight which decreases exponentially (division by 2) as the resolution decreases
        # this gives higher resolution outputs more weight in the loss
        weights = np.array([1 / (2 ** i) for i in range(len(deep_supervision_scales))])
        # we don't use the lowest 2 outputs. Normalize weights so that they sum to 1
        weights = weights / weights.sum()
        # now wrap the loss
        loss = DeepSupervisionWrapper(loss, weights)
        return loss
```
        
I don't think you can mix trainers while using the command but you can easily create new trainers by mixing the codes for example, the loss and epoch modification here. Add this class to one of your trainer variant files and you should be okay

**Comment by BSonya** [comment]
Thankyou for your response!
I have installed nnUNet using `pip install nnunet` as described in the documentation. To follow the solution described above, do I need to install the editable version of it using git clone .... ?

**Comment by BSonya** [comment]
Infact, I do have installed the modifiable version using
`git clone https://github.com/MIC-DKFZ/nnUNet.git`
`cd nnUNet`
`pip install -e .`
But the documentation doesn't describe the use of editable script files. Since I am a beginner, I would highly appreciate any further help.
Thanks

**Comment by MachtaYassine** [comment]
I'm guessing you're not using a code editor like vscode then ?

if you are, you can find the trainers in nnUNet/nnunetv2/training/nnUNetTrainer/variants and add the class i just displayed above.

if you are not using a code editor maybe you can try it from the Terminal you are using by going to the variants directory and using nano on the .py paths. I have never done that though so i don't know if it is recommended. so i recommend setting up a code editor. Sorry i couldn't be of further help

**Comment by BSonya** [comment]
Yes, I do have vs code. I just don't know the sequence in which I need to run the script files for planning, preprocessing & then training. Also, I wonder do I need to change directly every time I run a script file. 
But I'm trying to learn.
Thanks for the help!
